### PR TITLE
Make sure a subdataset is saved with a complete .gitmodules record

### LINKED ISSUE #6547: save -d .  submodule - doesn't add to .gitmodules
I have prepared a new http://github.com/datalad/talk-2022-abcd which was originally an rsync from the mila talk repo, committed new changes, created repo on github, pushed there and decided to save that new talk within the https://github.com/datalad/artwork dataset. But as a result -- there is a commit with subproject in git, but no changes to .gitmodules .  Not yet sure if it is result of my ad-hoc creation (rsync) of that repo at the beginning or something more  benign.

```shell
(git-annex)lena:~datalad/artwork[master]git
$> datalad save -m 'adding subdataset with abcd talk' -d . talks/2022-abcd/
add(ok): talks/2022-abcd (file)                                                                                                                   
save(ok): . (dataset)                                                                                                                             
action summary:                                                                                                                                   
  add (ok: 1)
  save (notneeded: 1, ok: 1)

$> git -C talks/2022-abcd/ log | head                                      
commit 438b95f361afa3fd4bacbfef3301bff6b83e8fee
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Mon Mar 14 09:09:58 2022 -0400

    just added a note to point to ABCD repronim course and YODA

commit 2f4408882374cc5ab7d0b4ab10f3e2bc81722665
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Mon Mar 14 09:03:46 2022 -0400

$> git status
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   logos/logo_wide.svg

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	logos/logo_solo.svg.2020_10_05_00_06_31.0.svg
	videos/Datalad Hirni webinar with Ben Poldrack (Dec 17, 2019)-380036896.mp4

$> git submodule
 96b79b97fec8f95bbb72b45286dd7119adb8ce36 talks/2019-mila (heads/master)
fatal: no submodule mapping found in .gitmodules for path 'talks/2022-abcd'

$> grep 2022 .gitmodules 

$> datalad --version
datalad 0.15.6+823.g9be2b482e

(git-annex)lena:~datalad/artwork[master]git
$> ls -dl talks/*/.git
drwx------ 1 yoh yoh 158 Mar 14 09:12 talks/2019-mila/.git/
drwx------ 1 yoh yoh 222 Mar 14 09:12 talks/2022-abcd/.git/
drwx------ 1 yoh yoh 158 Mar 14 09:12 talks/bernstein-2018-gofair/.git/
drwx------ 1 yoh yoh 158 Mar 14 09:12 talks/openneuro-2018/.git/
drwx------ 1 yoh yoh 176 Mar 14 09:12 talks/talk-fosdem-2020/.git/

$> git show -p    
commit 7087674916c762a214c6351057d7af8283e8afcf (HEAD -> master)
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Mon Mar 14 09:11:50 2022 -0400

    adding subdataset with abcd talk

diff --git a/talks/2022-abcd b/talks/2022-abcd
new file mode 160000
index 0000000..438b95f
--- /dev/null
+++ b/talks/2022-abcd
@@ -0,0 +1 @@
+Subproject commit 438b95f361afa3fd4bacbfef3301bff6b83e8fee

```

<details><summary>DataLad 0.15.6+823.g9be2b482e WTF (configuration, credentials, datalad, dataset, dependencies, environment, extensions, git-annex, location, metadata_extractors, metadata_indexers, python, system)</summary>

# WTF
## credentials 
  - keyring: 
    - active_backends: 
      - SecretService Keyring
      - PlaintextKeyring with no encyption v.1.0 at /home/yoh/.local/share/python_keyring/keyring_pass.cfg
    - config_file: /home/yoh/.config/python_keyring/keyringrc.cfg
    - data_root: /home/yoh/.local/share/python_keyring
## datalad 
  - version: 0.15.6+823.g9be2b482e
## dataset 
  - branches: 
    - git-annex@d7756ee
    - master@7087674
  - id: None
  - metadata: <SENSITIVE, report disabled by configuration>
  - path: /home/yoh/proj/datalad/artwork
  - repo: AnnexRepo
## dependencies 
  - annexremote: 1.5.0
  - boto: 2.49.0
  - cmd:7z: 16.02
  - cmd:annex: 10.20220222+git23-g51c528980-1~ndall+1
  - cmd:bundled-git: 2.30.2
  - cmd:git: 2.30.2
  - cmd:system-git: 2.34.1
  - cmd:system-ssh: 8.7p1
  - exifread: 2.3.2
  - humanize: 3.14.0
  - iso8601: 1.0.2
  - keyring: 23.5.0
  - keyrings.alt: 4.1.0
  - msgpack: 1.0.3
  - mutagen: 1.45.1
  - platformdirs: 2.5.0
  - requests: 2.27.1
## environment 
  - GIT_PAGER: less --no-init --quit-if-one-screen
  - LANG: en_US.UTF-8
  - PATH: /home/yoh/proj/datalad/datalad-master/venvs/dev3/bin:/home/yoh/gocode/bin:/home/yoh/gocode/bin:/home/yoh/bin:/home/yoh/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
## extensions 
  - deprecated: 
    - description: Deprecated functionality
    - entrypoints: 
      - datalad_deprecated.ls.Ls: 
        - class: Ls
        - load_error: None
        - module: datalad_deprecated.ls
        - names: 
    - load_error: None
    - module: datalad_deprecated
    - version: 0.1
## git-annex 
  - build flags: 
    - Assistant
    - Webapp
    - Pairing
    - Inotify
    - DBus
    - DesktopNotify
    - TorrentParser
    - MagicMime
    - Feeds
    - Testsuite
    - S3
    - WebDAV
  - dependency versions: 
    - aws-0.22
    - bloomfilter-2.0.1.0
    - cryptonite-0.26
    - DAV-1.3.4
    - feed-1.3.0.1
    - ghc-8.8.4
    - http-client-0.6.4.1
    - persistent-sqlite-2.10.6.2
    - torrent-10000.1.1
    - uuid-1.3.13
    - yesod-1.6.1.0
  - key/value backends: 
    - SHA256E
    - SHA256
    - SHA512E
    - SHA512
    - SHA224E
    - SHA224
    - SHA384E
    - SHA384
    - SHA3_256E
    - SHA3_256
    - SHA3_512E
    - SHA3_512
    - SHA3_224E
    - SHA3_224
    - SHA3_384E
    - SHA3_384
    - SKEIN256E
    - SKEIN256
    - SKEIN512E
    - SKEIN512
    - BLAKE2B256E
    - BLAKE2B256
    - BLAKE2B512E
    - BLAKE2B512
    - BLAKE2B160E
    - BLAKE2B160
    - BLAKE2B224E
    - BLAKE2B224
    - BLAKE2B384E
    - BLAKE2B384
    - BLAKE2BP512E
    - BLAKE2BP512
    - BLAKE2S256E
    - BLAKE2S256
    - BLAKE2S160E
    - BLAKE2S160
    - BLAKE2S224E
    - BLAKE2S224
    - BLAKE2SP256E
    - BLAKE2SP256
    - BLAKE2SP224E
    - BLAKE2SP224
    - SHA1E
    - SHA1
    - MD5E
    - MD5
    - WORM
    - URL
    - X*
  - local repository version: 8
  - operating system: linux x86_64
  - remote types: 
    - git
    - gcrypt
    - p2p
    - S3
    - bup
    - directory
    - rsync
    - web
    - bittorrent
    - webdav
    - adb
    - tahoe
    - glacier
    - ddar
    - git-lfs
    - httpalso
    - borg
    - hook
    - external
  - supported repository versions: 
    - 8
    - 9
    - 10
  - upgrade supported from repository versions: 
    - 0
    - 1
    - 2
    - 3
    - 4
    - 5
    - 6
    - 7
    - 8
    - 9
    - 10
  - version: 10.20220222+git23-g51c528980-1~ndall+1
## location 
  - path: /home/yoh/proj/datalad/artwork
  - type: dataset
## metadata_extractors 
  - annex (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.annex
    - version: None
  - audio (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.audio
    - version: None
  - datacite (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.datacite
    - version: None
  - datalad_core (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.datalad_core
    - version: None
  - datalad_rfc822 (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.datalad_rfc822
    - version: None
  - exif (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.exif
    - version: None
  - frictionless_datapackage (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.frictionless_datapackage
    - version: None
  - image (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.image
    - version: None
  - xmp (datalad 0.15.4+736.gecd457775): 
    - distribution: datalad 0.15.4+736.gecd457775
    - load_error: None
    - module: datalad.metadata.extractors.xmp
    - version: None
## python 
  - implementation: CPython
  - version: 3.9.10
## system 
  - distribution: debian/testing/bookworm
  - encoding: 
    - default: utf-8
    - filesystem: utf-8
    - locale.prefered: UTF-8
  - max_path_length: 286
  - name: Linux
  - release: 5.15.0-3-amd64
  - type: posix
  - version: #1 SMP Debian 5.15.15-2 (2022-01-30)
</details>

### PR DESCRIPTION
**Pull request by mih**
Previously, the test for adding a gitmodules record was incomplete
and relied on the assumption that for a new subdataset the type
`dataset` was not determined by the initial `status()` inspection.
However, this is not the case, when an explicit subdataset path
with a trailing path-separator is given to `save()`.

This change amends the test condition to also test for the state
being `untracked`. In principle this alone should be a sufficient
condition. However, due to the complexity of the decision making
in the present implementation, the previous condition as also
kept as an alternative, in order to prevent unexpected side-effect.

Fixes datalad/datalad#6547

#### 🐛 Bug Fixes
- Ensure saving a new subdataset to a superdataset yields a valid `.gitmodules` record regardless of whether and how a path constraint is given to the `save()` call. Fixes #6547

### CODE REVIEW

**Comment by mih** [comment]
Test failure is real, investigating...

**Comment by mih** [comment]
The changes in this PR are now a precondition to an upcoming PR addressing #6791 -- depending on the timing of review, I will close this PR and integrate it into a substantially larger update that introduces typechange detection.

File: datalad/core/local/save.py [hunk 1]
```diff
@@ -279,14 +279,16 @@ def __call__(path=None,
...
+                        superds_status[subds_path] = superds_status.get(
+                            subds_path,
+                            # if we got nothing yet:
                             # shot from the hip, some status config
                             # to trigger this specific super/sub
                             # relation to be saved
-                            state='untracked',
-                            type='dataset')
+                            dict(state='untracked', type='dataset')
+                        )
```
Comment (yarikoptic) [inline datalad/core/local/save.py:291]: smells like we would benefit from more type annotations and adding type checking

**Comment by yarikoptic** [comment]
seems to fix a bug and has a test and all green -- what could stop us? ;)  I will CP to maint within yet another PR

**Comment by yarikoptic** [comment]
FWIW -- decided not to bother with a PR, CPed, ran the test, pushed with tip to 6ec85d4d51af9997f763c9bd9f989fb0fd78fe55

**Pull request merged**